Add Generated Columns support for MySQL 5.7

kamipo opened the pull request:
Generated Columns is a new feature available in MySQL 5.7.5 and up.
See https://dev.mysql.com/doc/refman/5.7/en/create-table.html.

Example:

``` ruby
create_table :generated_columns do |t|
  t.string  :name
  t.virtual :upper_name,  type: :string,  as: 'UPPER(name)'
  t.virtual :name_length, type: :integer, as: 'LENGTH(name)', stored: true
end
```

kamipo committed:
Add Generated Columns support for MySQL 5.7

Generated Columns is a new feature available in MySQL 5.7.5 and up.
See https://dev.mysql.com/doc/refman/5.7/en/create-table.html.

Example:

    create_table :generated_columns do |t|
      t.string  :name
      t.virtual :upper_name,  type: :string,  as: "UPPER(name)"
      t.virtual :name_length, type: :integer, as: "LENGTH(name)", stored: true
    end

nonmadden commented on activerecord/test/cases/adapters/mysql2/generated_column_test.rb:
```diff
@@ -0,0 +1,58 @@
+    assert_match(/t\.string\s+"upper_name",\s+limit: 255,\s+as: "UPPER\(name\)", virtual: true$/, output)
+    assert_match(/t\.integer\s+"name_length",\s+limit: 4,\s+as: "LENGTH\(name\)", stored: true$/, output)
+  end
+end
```
`end if ActiveRecord::Base.connection.supports_generated_columns?`

jeremy commented:
Nice work!

Oracle, SQL Server, MariaDB, and others support virtual/computed/generated columns too. "Generated" seems to be the least common name: https://en.wikipedia.org/wiki/Virtual_column

Note that all the other column options don't make sense for a virtual column. Perhaps they should be explicitly declared as virtual columns rather than adding a handful of options to normal columns:
```diff
 change_table :people do |t|
-  t.string :lower_name, as: 'LOWER(name)'
-  t.virtual :lower_name, as: 'LOWER(name)'
 end
```

kamipo commented:
I added `virtual` method for declaring virtual columns such like oracle adapter.
https://github.com/rsim/oracle-enhanced/blob/3c42131db82b64ac41645db3affc6e4650289df6/spec/active_record/connection_adapters/oracle_enhanced_schema_dump_spec.rb#L370-L388

```ruby
create_table :generated_columns do |t|
  t.string  :name
  t.virtual :upper_name,  type: :string,  as: 'UPPER(name)'
  t.virtual :name_length, type: :integer, as: 'LENGTH(name)', stored: true
end
```

And added `AbstractAdapter#supports_virtual_columns?` and renamed `s/generated/virtual/`.

@jeremy What do you think about it?

matthewd commented:
This seems awfully view-like. Is it DSL-worthy?

We generally take the position that if you need more advanced database features, you're better served by `structure.sql`.

rafaelfranca commented:
r? @jeremy

jeremy commented:
@kamipo I hate to ask you so late… but could you rebase master? 😊

kamipo commented:
Rebased on master!

prathamesh-sonpatki commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,17 @@
+        create_table :generated_columns do |t|
+          t.string  :name
+          t.virtual :upper_name,  type: :string,  as: 'UPPER(name)'
+          t.virtual :name_length, type: :integer, as: 'LENGTH(name)', virtual: :stored
```
`t.virtual :name_length, type: :integer, as: 'LENGTH(name)', virtual: :stored` 

Does it look odd having `t.virtual` and then again `virtual: :stored`? May be `strategy: :stored`?

prathamesh-sonpatki commented on activerecord/test/cases/adapters/mysql2/generated_column_test.rb:
```diff
@@ -0,0 +1,58 @@
+require 'support/schema_dumping_helper'
+
+if ActiveRecord::Base.connection.supports_virtual_columns?
+class Mysql2GeneratedColumnTest < ActiveRecord::Mysql2TestCase
```
Indentation needs to be fixed.

yahonda commented:
I am interested in this pull request.

Oracle enhanced adapter already implemented this feature (now it needs a lot of refactoring... anyway), I'd be happy if ActiveRecord itself implements this feature.

Here is one request / proposal for this `virtual: stored` option. 

```ruby
t.virtual :name_length, type: :integer, as: "LENGTH(name)", virtual: :stored
```

How about changing it to `stored: true` by not using "virtual" keyword here.

Because the current pull request uses "VIRTUAL" keyword for two meanings. 

- if this column is generated from other columns and/or values (generic)
- Storing this column to disk or not (MySQL specific)

In MySQL this feature is called "Generated Columns", then using "VIRTUAL" keyword to decide store generated column is straightforward. Once changing "Generated Columns" to "Virtual Columns", I was little bit confused why there is "VIRTUAL" option here.

Accoring to [14.1.18.7 CREATE TABLE and Generated Columns](https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html) VIRTUAL or STORED are exclusive option and VIRTUAL is the default value if I understand it correctly.
```
The VIRTUAL or STORED keyword indicates how column values are stored, which has implications for column use:

- VIRTUAL: Column values are not stored, but are evaluated when rows are read, immediately after any BEFORE triggers. A virtual column takes no storage.

InnoDB supports secondary indexes on virtual columns. See Section 14.1.18.8, “Secondary Indexes and Generated Columns”.

- STORED: Column values are evaluated and stored when rows are inserted or updated. A stored column does require storage space and can be indexed.

The default is VIRTUAL if neither keyword is specified.
```
It would be appreciated if request / proposal is considered.

jeremy commented:
Agree this is a boolean choice for whether the column value is stored. The Wikipedia page and other database docs use "persisted" and "stored" language as well. Strong reason to use a boolean keyword arg.

kamipo committed:
Clear stub_version value

kamipo commented:
Thank you for your suggestion! I changed to `stored: true`.

Originally first implementation was here:

```ruby
create_table :generated_columns do |t|
  t.string  :name
  t.string  :upper_name,  as: 'UPPER(name)',  virtual: true
  t.integer :name_length, as: 'LENGTH(name)', virtual: :stored
end
```

Generated Column cannot omit `type`, so I implemented `virtual` feature as an option.
Another reason, in Rails 5.0, couldn't dump `type` option. https://github.com/rails/rails/blob/v5.0.1/activerecord/lib/active_record/connection_adapters/abstract/schema_dumper.rb#L12

But now, `type` option is not overwritten. https://github.com/rails/rails/blob/f543b97fcc52168119d4e016eb50afbda702e601/activerecord/lib/active_record/connection_adapters/abstract/schema_dumper.rb#L10

Current implementation in the PR is here:

```ruby
create_table :generated_columns do |t|
  t.string  :name
  t.virtual :upper_name,  type: :string,  as: 'UPPER(name)'
  t.virtual :name_length, type: :integer, as: 'LENGTH(name)', stored: true
end
```

Thank you again!

yahonda commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -141,6 +141,10 @@ def supports_datetime_with_precision?
       end
 
+      def supports_virtual_columns?
+        !mariadb? && version >= "5.7.5"
```
I do not have MariaDB environment yet, looks like it always returns false with MariaDB.  MariaDB 5.2 supports virtual columns and this pull request is considering it by implementing PERSISTENT option.  Forgive me if I misread how ! works...

kamipo replied:
`supports_virtual_columns?` is only used in schema **dumping**.
https://github.com/rails/rails/pull/22589/files#diff-621ba9ae39e5952c32d9f06ecf8eb433R63
https://github.com/rails/rails/pull/22589/files#diff-f50ea2a3acc95dd2652e9cb1b3f75961R18

It is because schema creation support is easy, but schema dumping support is not easy. Current implementation (extracting `as` expression) is depend on `information_schema`, but I couldn't find the information in MariaDB at least 10.1.

So I dicided to only support schema creation for MariaDB (does not support schema dumping yet now).

yahonda replied:
Thanks for the information. I should have read entire changes. 

Tested with MariaDB 10.2.2 beta and also I am unable to find virtual column expressions such as `'UPPER(name)'` from `information_schema.columns`.

```sql
MariaDB [yahonda]> CREATE TABLE `virtual_columns` (`id` bigint AUTO_INCREMENT PRIMARY KEY, `name` varchar(255), `upper_name` varchar(255) AS (UPPER(name)), `name_length` int AS (LENGTH(name)) PERSISTENT) ENGINE=InnoDB;
Query OK, 0 rows affected (0.05 sec)

MariaDB [yahonda]> show create table virtual_columns\G
*************************** 1. row ***************************
       Table: virtual_columns
Create Table: CREATE TABLE `virtual_columns` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) DEFAULT NULL,
  `upper_name` varchar(255) AS (UPPER(name)) VIRTUAL,
  `name_length` int(11) AS (LENGTH(name)) PERSISTENT,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1
1 row in set (0.00 sec)

MariaDB [yahonda]> select * from information_schema.columns
    -> where table_name = 'virtual_columns' and column_name in ('upper_name', 'name_length')\G
*************************** 1. row ***************************
           TABLE_CATALOG: def
            TABLE_SCHEMA: yahonda
              TABLE_NAME: virtual_columns
             COLUMN_NAME: upper_name
        ORDINAL_POSITION: 3
          COLUMN_DEFAULT: NULL
             IS_NULLABLE: YES
               DATA_TYPE: varchar
CHARACTER_MAXIMUM_LENGTH: 255
  CHARACTER_OCTET_LENGTH: 255
       NUMERIC_PRECISION: NULL
           NUMERIC_SCALE: NULL
      DATETIME_PRECISION: NULL
      CHARACTER_SET_NAME: latin1
          COLLATION_NAME: latin1_swedish_ci
             COLUMN_TYPE: varchar(255)
              COLUMN_KEY: 
                   EXTRA: VIRTUAL
              PRIVILEGES: select,insert,update,references
          COLUMN_COMMENT: 
*************************** 2. row ***************************
           TABLE_CATALOG: def
            TABLE_SCHEMA: yahonda
              TABLE_NAME: virtual_columns
             COLUMN_NAME: name_length
        ORDINAL_POSITION: 4
          COLUMN_DEFAULT: NULL
             IS_NULLABLE: YES
               DATA_TYPE: int
CHARACTER_MAXIMUM_LENGTH: NULL
  CHARACTER_OCTET_LENGTH: NULL
       NUMERIC_PRECISION: 10
           NUMERIC_SCALE: 0
      DATETIME_PRECISION: NULL
      CHARACTER_SET_NAME: NULL
          COLLATION_NAME: NULL
             COLUMN_TYPE: int(11)
              COLUMN_KEY: 
                   EXTRA: PERSISTENT
              PRIVILEGES: select,insert,update,references
          COLUMN_COMMENT: 
2 rows in set (0.00 sec)

MariaDB [yahonda]> select version();
+----------------+
| version()      |
+----------------+
| 10.2.2-MariaDB |
+----------------+
1 row in set (0.00 sec)

MariaDB [yahonda]>  
```

I understand the schema dumper implementation difficulties with MariaDB, still I think we need to have another name to switch schema dumping virtual column expressions or not.

Since it will be confusing users (like me) can create virtual columns even if `supports_virtual_columns?` returns `false`.  

How about `supports_dump_virtual_column_expressions?` ?

kamipo replied:
How about adding the comment (partially support for MariaDB) like `supports_index_sort_order?`?
https://github.com/rails/rails/blob/v5.0.1/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb#L110-L114

I'm not sure there are similar DBs other than MariaDB.

yahonda replied:
@kamipo

Thanks for the suggestion. I still prefer my idea but I do not have strong opinion to handle this 
by adding comments at the `supports_virtual_columns?` then I'd like to hear from Jeremy.

@jeremy

I'd like to get some opinion from MariaDB users (I think you are the one of them) 
are happy to get somewhat limited virtual column support - migration is supported but schema dumper is not supported. Or prefer to wait until MariaDB information_schema can generate virtual column expressions like MySQL 5.7 does.

Thanks guys.

jeremy replied:
I use MySQL locally and MariaDB in production, so this setup would work for me.

I think it is pretty confusing for people doing development against MariaDB locally, though. Running tests dumps dev schema to load for the test db, so your test runs would fail with a schema you can't round-trip :(

People could still use SQL dumps for schema and use this declarative style for migrations, however.

What route do you think is friendliest and most useful to people doing local development with MariaDB?

yahonda replied:
Thanks for the reply. Oh, I did not expect that combination.

One idea is not supporting virtual columns schema migrations at all. Well, it does not look friendliest but at least it is somewhat kind since users get to know the virtual column is not supported from the beginning, not at the timing when they are performing schema dumps after creating a bunch of migrations including virtual columns.

Another one is adding comments or documents to use SQL dump for virtual columns. I do not have other ideas yet.

jeremy replied:
> One idea is not supporting virtual columns schema migrations at all

Agree this sets the clearest expectations which is probably the friendliest choice. 👍

yahonda replied:
Glad to know you agreed.  Let's support virtual columns schema migrations and schema dump only for MySQL 5.7 or higher version, no MariaDB support added this time.

yahonda commented on activerecord/test/cases/adapters/mysql2/virtual_column_test.rb:
```diff
@@ -0,0 +1,59 @@
+      assert_equal "rails", VirtualColumn.take.lower_name
+    end
+
+    def test_schema_dumping
```
This test fails with these two MySQL servers.

- Server version: 8.0.0-dmr MySQL Community Server (GPL)
- Server version: 5.7.17 MySQL Community Server (GPL)

```ruby
$ ARCONN=mysql2 bundle exec ruby -Itest test/cases/adap                                        ters/mysql2/virtual_column_test.rb -n test_schema_dumping

Using mysql2
Run options: -n test_schema_dumping --seed 61883

# Running:

F

Finished in 0.835548s, 1.1968 runs/s, 2.3936 assertions/s.

  1) Failure:
Mysql2VirtualColumnTest#test_schema_dumping [test/cases/adapters/mysql2/virtual_column_test.rb:55]:
Expected /t\.virtual\s+"upper_name",\s+type: :string,\s+as: "UPPER\(name\)"$/ to match "# This file is auto-generated from the current state of the database. Instead\n# of editing this file, please use the migrations feature of Active Record to\n# incrementally modify your database, and then regenerate this schema definition.\n#\n# Note that this schema.rb definition is the authoritative source for your\n# database schema. If you need to create the application database on another\n# system, you should be using db:schema:load, not running all the migrations\n# from scratch. The latter is a flawed and unsustainable approach (the more migrations\n# you'll amass, the slower it'll run and the greater likelihood for issues).\n#\n# It's strongly recommended that you check this file into your version control system.\n\nActiveRecord::Schema.define(version: 0) do\n\n  create_table \"virtual_columns\", force: :cascade, options: \"ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci\" do |t|\n    t.string \"name\"\n    t.virtual \"upper_name\", type: :string, as: \"upper(`name`)\"\n    t.virtual \"name_length\", type: :integer, as: \"length(`name`)\", stored: true\n  end\n\nend\n".

1 runs, 2 assertions, 1 failures, 0 errors, 0 skips
$
```

It is likely due to UPPER and LENGTH function are converted into lowercase upper and length at MySQL server side.

```sql
mysql> CREATE TABLE `virtual_columns` (`id` bigint AUTO_INCREMENT PRIMARY KEY, `name` varchar(255), `upper_name` varchar(255) AS (UPPER(name)), `name_length` int AS (LENGTH(name)) STORED) ENGINE=InnoDB;
Query OK, 0 rows affected (0.00 sec)

mysql> show create table `virtual_columns`\G
*************************** 1. row ***************************
       Table: virtual_columns
Create Table: CREATE TABLE `virtual_columns` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) DEFAULT NULL,
  `upper_name` varchar(255) GENERATED ALWAYS AS (upper(`name`)) VIRTUAL,
  `name_length` int(11) GENERATED ALWAYS AS (length(`name`)) STORED,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1
1 row in set (0.00 sec)

mysql>
```

kamipo replied:
ah... I'll change tests to ignore case...

kamipo replied:
Changed in https://github.com/rails/rails/pull/22589/commits/6e76322207510662ccaea8f4e0d3e59dd806ff39.

yahonda commented:
Thanks for updating the pull request and explaining background changes. The new syntax is great.
Tested at my environment and added two comments.

Thanks.

kamipo committed:
Ignore case `test_schema_dumping` in `Mysql2VirtualColumnTest`

In MySQL 5.7.17 and 8.0.0-dmr, functions in expressions appears as lower
case (previously upper case). So it should be handled as ignore case.

kamipo committed:
Add Virtual Columns support for MariaDB

kamipo commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -141,6 +141,14 @@ def supports_datetime_with_precision?
+        if mariadb?
+          version >= "5.2.0"
+        else
+          version >= "5.7.5"
```
I implemented extracting the expression for MariaDB in e0841ac6aaa873a7125342d4465c28e00dd919e3. Thanks!

yahonda commented:
Thanks @kamipo

Jeremy and I were discussing just supporting virtual columns for MySQL 5.7 since MariaDB `information_schema.columns` does not provide virtual columns expressions, then it will not support schema dump.

Instead of dropping MariaDB support, this https://github.com/rails/rails/commit/e0841ac6aaa873a7125342d4465c28e00dd919e3
supports virtual columns for MariaDB by using `create_table_info` method which extracts `show create table` output and retrieve necessary virtual columns expression information.

As long as schema dump can be generated by information_schema for MySQL, by extracting `show table` output for MariaDB I'm :+1: for this pull request.

rafaelfranca commented on activerecord/test/cases/adapters/mysql2/virtual_column_test.rb:
```diff
@@ -0,0 +1,59 @@
+
+    def test_virtual_column
+      column = VirtualColumn.columns_hash["upper_name"]
+      assert column.virtual?
```
`assert_predicate`

yahonda commented on activerecord/test/cases/adapters/mysql2/virtual_column_test.rb:
```diff
@@ -0,0 +1,59 @@
+      assert_equal "rails", VirtualColumn.take.lower_name
+    end
+
+    def test_schema_dumping
```
It still fails with MySQL 8.0.0.

```ruby
$ ARCONN=mysql2 bundle exec ruby -w -Itest test/cases/adapters/mysql2/virtual_column_test.rb
/home/yahonda/git/rails/activesupport/lib/active_support/core_ext/enumerable.rb:20: warning: method redefined; discarding old sum
Using mysql2
Run options: --seed 3987

# Running:

...F

Finished in 1.154189s, 3.4656 runs/s, 12.1297 assertions/s.

  1) Failure:
Mysql2VirtualColumnTest#test_schema_dumping [test/cases/adapters/mysql2/virtual_column_test.rb:55]:
Expected /t\.virtual\s+"upper_name",\s+type: :string,\s+as: "UPPER\(name\)"$/i to match "# This file is auto-generated from the current state of the database. Instead\n# of editing this file, please use the migrations feature of Active Record to\n# incrementally modify your database, and then regenerate this schema definition.\n#\n# Note that this schema.rb definition is the authoritative source for your\n# database schema. If you need to create the application database on another\n# system, you should be using db:schema:load, not running all the migrations\n# from scratch. The latter is a flawed and unsustainable approach (the more migrations\n# you'll amass, the slower it'll run and the greater likelihood for issues).\n#\n# It's strongly recommended that you check this file into your version control system.\n\nActiveRecord::Schema.define(version: 0) do\n\n  create_table \"virtual_columns\", force: :cascade, options: \"ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci\" do |t|\n    t.string \"name\"\n    t.virtual \"upper_name\", type: :string, as: \"upper(`name`)\"\n    t.virtual \"name_length\", type: :integer, as: \"length(`name`)\", stored: true\n  end\n\nend\n".

4 runs, 14 assertions, 1 failures, 0 errors, 0 skips
```

Here "name" inside the upper and length function needs backquote.

kamipo replied:
I confirmed MySQL 8.0.0 always quotes column names. So I'll change the definition in the tests includes back quotes. Thanks!

```
root@localhost [test] > CREATE TABLE `virtual_columns` (`id` bigint AUTO_INCREMENT PRIMARY KEY, `name` varchar(255), `upper_name` varchar(255) AS (UPPER(name)), `name_length` int AS (LENGTH(name)) STORED) ENGINE=InnoDB;
Query OK, 0 rows affected (0.01 sec)

root@localhost [test] > show create table virtual_columns\G                                                                                 *************************** 1. row ***************************
       Table: virtual_columns
Create Table: CREATE TABLE `virtual_columns` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) DEFAULT NULL,
  `upper_name` varchar(255) GENERATED ALWAYS AS (upper(`name`)) VIRTUAL,
  `name_length` int(11) GENERATED ALWAYS AS (length(`name`)) STORED,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
1 row in set (0.00 sec)

root@localhost [test] > select column_name, column_type, generation_expression from information_schema.columns where table_schema = database() and table_name = 'virtual_columns';
+-------------+--------------+-----------------------+
| COLUMN_NAME | COLUMN_TYPE  | GENERATION_EXPRESSION |
+-------------+--------------+-----------------------+
| id          | bigint(20)   |                       |
| name        | varchar(255) |                       |
| name_length | int(11)      | length(`name`)        |
| upper_name  | varchar(255) | upper(`name`)         |
+-------------+--------------+-----------------------+
4 rows in set (0.00 sec)

root@localhost [test] > select @@version;
+---------------+
| @@version     |
+---------------+
| 8.0.0-dmr-log |
+---------------+
1 row in set (0.00 sec)
```

kamipo replied:
Fixed in 7b291ff.

yahonda replied:
Thanks I also confirmed it has been fixed.

kamipo committed:
Prefer `assert_predicate` than `assert`

kamipo committed:
The definition in the tests includes backquotes for MySQL 8.0.0

I confirmed MySQL 8.0.0 always quotes column names. So I'll change the
definition in the tests includes backquotes.

```
root@localhost [test] > CREATE TABLE `virtual_columns` (`id` bigint AUTO_INCREMENT PRIMARY KEY, `name` varchar(255), `upper_name` varchar(255) AS (UPPER(name)), `name_length` int AS (LENGTH(name)) STORED) ENGINE=InnoDB;
Query OK, 0 rows affected (0.01 sec)

root@localhost [test] > show create table virtual_columns\G                                                                                 *************************** 1. row ***************************
       Table: virtual_columns
Create Table: CREATE TABLE `virtual_columns` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) DEFAULT NULL,
  `upper_name` varchar(255) GENERATED ALWAYS AS (upper(`name`)) VIRTUAL,
  `name_length` int(11) GENERATED ALWAYS AS (length(`name`)) STORED,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
1 row in set (0.00 sec)

root@localhost [test] > select column_name, column_type, generation_expression from information_schema.columns where table_schema = database() and table_name = 'virtual_columns';
+-------------+--------------+-----------------------+
| COLUMN_NAME | COLUMN_TYPE  | GENERATION_EXPRESSION |
+-------------+--------------+-----------------------+
| id          | bigint(20)   |                       |
| name        | varchar(255) |                       |
| name_length | int(11)      | length(`name`)        |
| upper_name  | varchar(255) | upper(`name`)         |
+-------------+--------------+-----------------------+
4 rows in set (0.00 sec)

root@localhost [test] > select @@version;
+---------------+
| @@version     |
+---------------+
| 8.0.0-dmr-log |
+---------------+
1 row in set (0.00 sec)
```

jeremy commented:
Merged @ 65bf1c60053e727835e06392d27a2fb49665484c. Thank you @kamipo!!
